runelite-client/src/main/java/net/runelite/client/plugins/farmingtracker/FarmingTrackerPanel.java

```diff
@@ -0,0 +1,289 @@
+			};
+			icon.onChanged(resize);
+			resize.run();
+
+			if (config.patch() == tab)
+			{
+				tabs.setSelectedComponent(scroller);
+			}
+			tabs.addChangeListener(e ->
+			{
+				if (tabs.getSelectedComponent() == scroller)
+				{
+					config.setPatch(tab);
+				}
+			});
+		});
+		add(tabs, BorderLayout.CENTER);
+	}
+
+	void update()
+	{
+		long unixNow = Instant.now().getEpochSecond();
+		log.debug("Updating panel with username {}", client.getUsername());
+		for (FarmingPatchPanel panel : patchPanels)
+		{
+			FarmingPatch patch = panel.getPatch();
+			String group = FarmingTrackerConfig.KEY_NAME + "." + client.getUsername() + "." + patch.getRegion().getRegionID();
+			String key = Integer.toString(patch.getVarbit().getId());
+			String storedValue = configManager.getConfiguration(group, key);
+			long unixTime = 0;
+			int value = 0;
+			if (storedValue != null)
+			{
+				String[] parts = storedValue.split(":");
+				if (parts.length == 2)
+				{
+					try
+					{
+						value = Integer.parseInt(parts[0]);
+						unixTime = Long.parseLong(parts[1]);
+					}
+					catch (NumberFormatException e)
+					{
+					}
+				}
+			}
+
+			PatchState state = unixTime <= 0 ? null : patch.getImplementation().forVarbitValue(value);
+			if (state == null)
+			{
+				panel.getIcon().setIcon(null);
+				panel.getIcon().setToolTipText("Unknown state");
+				panel.getProgress().setMaximum(0);
+				panel.getProgress().setValue(0);
+				panel.getEstimate().setText("Unknown");
+			}
+			else
+			{
+				if (state.getProduce().getItemID() < 0)
+				{
+					panel.getIcon().setIcon(null);
+					panel.getIcon().setToolTipText("Unknown state");
+				}
+				else
+				{
+					itemManager.getImage(state.getProduce().getItemID()).addTo(panel.getIcon());
+					panel.getIcon().setToolTipText(state.getProduce().getName());
+				}
+
+				int stage = state.getStage();
+				int stages = state.getCropState() == CropState.HARVESTABLE ?
+					state.getProduce().getHarvestStages() :
+					state.getProduce().getStages();
+				int tickrate = 0;
+				switch (state.getCropState())
+				{
+					case HARVESTABLE:
+						tickrate = state.getProduce().getRegrowTickrate() * 60;
+						break;
+					case GROWING:
+						tickrate = state.getProduce().getTickrate() * 60;
+						break;
+				}
+				if (tickrate > 0)
+				{
+					long tickNow = unixNow / tickrate;
+					long tickTime = unixTime / tickrate;
+					int delta = (int) (tickNow - tickTime);
+
+					long doneEstimate = ((stages - 1 - stage) + tickTime) * tickrate;
+
+					stage += delta;
+					if (stage >= stages)
+					{
+						stage = stages - 1;
+					}
+
+					if (doneEstimate < unixNow)
+					{
+						panel.getEstimate().setText("Done");
+					}
+					else if (config.estimateRelative())
+					{
+						int remaining = (int) (59 + doneEstimate - unixNow) / 60;
+						StringBuilder f = new StringBuilder();
+						f.append("Done in ");
+						int min = remaining % 60;
```

Comment:
There must be some date fromatting function ready to use, no neet to reinvent the wheel here

Reply:
I couldn't find a implementation in the standard library that would elide zero-valued segments

Reply:
I see